Add support for optional per-user mods in multiplayer (aka freemod)

smoogipoo opened the pull request:
Prereqs:
- [x] https://github.com/ppy/osu/pull/11640
- [x] https://github.com/ppy/osu/pull/11667
- [x] https://github.com/ppy/osu/pull/11666

I've used the terminology "freemods" when referring to the selection by the host (i.e. in song select), "allowedmods" when referring to the mods in API code (to follow the existing osu-web API), and "user mods" when referring to mods set by/affecting only the local user.

* A big chunk of this is refactoring of `ModSelectOverlay` to allow just enough customisability for 1) removing stacking, 2) allowing isvalid to be mutable, 3) adding a checkbox in the mod section headers
* Another big chunk is refactoring of the online play song selection. It was currently implemented in two places: `MatchSongSelect` (which has been renamed to `PlaylistsSongSelect`) and `MultiplayerMatchSongSelect`, while both have identical functionality besides playlists. All custom logic goes through the `SelectItem()` method now, rather than `OnStart()`.
* The rest should be fairly straight forward and easy to review - a few changes in classes like `ModIcon` which fixes colourisation inconsistencies and `ModDisplay` which adds a toggle for the initial expansion, just to improve the display.

Preview: https://drive.google.com/file/d/1gwA5A-WzdPcO22g2VvYvcI1anY9UcEzV/view
(ignore miscoloured mod icons - fixed above as mentioned)

smoogipoo committed:
Add mod validation utilities

smoogipoo committed:
Make ModSections overrideable

smoogipoo committed:
Fix type not being set

smoogipoo committed:
Abstractify ModSelectOverlay

smoogipoo committed:
Allow footer buttons to be customised

smoogipoo committed:
Initial implementation of freemod selection overlay

smoogipoo committed:
Actually populate the playlist item

smoogipoo committed:
Add AllowedMods to MultiplayerRoomSettings model

smoogipoo committed:
Fix possible nullref

smoogipoo committed:
Add mods to footer

smoogipoo committed:
Pass allowed mods and consume on server callback

smoogipoo committed:
Move ModSelectOverlay.IsValidMod to a property

smoogipoo committed:
Allow unstacking mods

smoogipoo committed:
Disallow selecting DT/HT/WU/WD as allowable freemods

smoogipoo committed:
Add initial UI for selecting extra mods

smoogipoo committed:
Update mods when resuming room subscreen

smoogipoo committed:
Move common song select implementation for online play

smoogipoo committed:
Fix allowed mods being copied into required mods

smoogipoo committed:
Once again disallow DT/etc as allowable mods

smoogipoo committed:
Add user and panel states

smoogipoo committed:
Extra mods -> user mods

smoogipoo committed:
Renamespace

smoogipoo committed:
Disable customisation of freemods, move stacking to property

smoogipoo committed:
Stack freemods for the local user

smoogipoo committed:
Reduce mod selection height

smoogipoo committed:
Pin mod position in participant panels

smoogipoo committed:
General cleanup

smoogipoo committed:
Disallow local user mod customisation

smoogipoo committed:
Fix mod icons potentially having incorrect colours

smoogipoo committed:
Merge branch 'master' into freemods

smoogipoo committed:
Remove usage of removed method

smoogipoo committed:
Fix unset key

smoogipoo committed:
Renamespace/rename MatchSongSelect -> PlaylistsSongSelect

frenzibyte commented:
This looks to be going in a circular dependency, https://github.com/ppy/osu-server-spectator/pull/50 depends on this PR for server methods, and this PR needs https://github.com/ppy/osu-server-spectator/pull/50 to be merged and deployed for working properly, otherwise method not found exceptions will occur.

peppy commented:
@frenzibyte that's just how clients and servers works. this will be the case for every client and server change, unless the common components are moved to a third project.

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs:
```diff
@@ -128,10 +135,39 @@ private void load()
                                                             Padding = new MarginPadding { Horizontal = 5 },
-                                                            Children = new Drawable[]
+                                                            Spacing = new Vector2(0, 10),
+                                                            Children = new[]
```
The contents here jump around a lot - part of it is because the mod display is in between the header and the button, but another part of it is because the parenting container is `Centre`-anchored; would probably make it `TopCentre` for a little less jumping around.

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs:
```diff
@@ -274,5 +352,10 @@ protected override void Dispose(bool isDisposing)
+
+        private class UserModSelectOverlay : ModSelectOverlay
+        {
+            protected override bool AllowCustomisation => false;
```
Is this correct? Does this mean that free mods cannot be customised by anyone (neither the host or each individual player)? I imagine that turns mods like difficulty adjust quite useless.

smoogipoo replied:
Again, this is kinda temporary because it's hard to work in. From the guest's perspective, it requires somehow clamping mod settings based on what the host picks, which I don't know if it's even possible with the current structure of things.

From the host's perspective, it requires handling of mod settings to propagate to the server, and a bit more work on `APIMod`'s equality comparer. I think this is better left to a further PR.

bdach commented:
The above is mostly a rudimentary first pass. This PR does a lot (too much?), so I'm sure I missed things.

On another note, the various mod overlays should really get varying header/description text.

smoogipoo commented on osu.Game/Online/Multiplayer/MultiplayerRoomSettings.cs:
```diff
@@ -32,13 +32,22 @@ public class MultiplayerRoomSettings : IEquatable<MultiplayerRoomSettings>
         [Key(4)]
         public IEnumerable<APIMod> Mods { get; set; } = Enumerable.Empty<APIMod>();
```
The breaking change part of that is the reason I didn't do it. But I agree otherwise.

smoogipoo committed:
Rename test

smoogipoo committed:
Make checkbox also respond to all mods selected

smoogipoo committed:
Exclude more mods from multiplayer

smoogipoo committed:
Refactor condition

This won't make any noticeable difference, but is the more correct way
to handle MultiMod because flattening works through infinite recursion levels.

smoogipoo commented on osu.Game/Overlays/Mods/ModSelectOverlay.cs:
```diff
@@ -39,6 +39,16 @@ public class ModSelectOverlay : WaveOverlayContainer
 
         protected readonly OsuSpriteText MultiplierLabel;
 
+        /// <summary>
+        /// Whether to allow customisation of mod settings.
+        /// </summary>
+        protected virtual bool AllowCustomisation => true;
```
Will also disable `DifficultyAdjustMod` (and all mods requiring config) for now. This is a temporary thing, with probably the biggest thing to handle being the changing of settings in order to propagate to server (i.e. binding to all settings, and making `APIMod` consider setting values in its equality comparer).

smoogipoo committed:
Hide customise button + multiplier label

smoogipoo committed:
Add option to select all

smoogipoo committed:
Refactor ModDisplay flag usage

smoogipoo committed:
Make ModIcon a bit more safe

smoogipoo commented on osu.Game/Screens/OnlinePlay/Match/FooterButtonFreeMods.cs:
```diff
@@ -0,0 +1,63 @@
+
+        public FooterButtonFreeMods()
+        {
+            ButtonContentContainer.Add(modDisplay = new ModDisplay
```
I... have no idea. It seems to work in the participant panel without `ExpandOnAppear = false` too, so I'll just revert this change.
It definitely didn't work in some case though o_O

smoogipoo committed:
Revert changes to ModDisplay

smoogipoo committed:
Add xmldoc

smoogipoo committed:
Clear freemods when ruleset is changed

peppy commented:
Would have been nice to split out the mod overlay changes (especially those with no behaviour change, like the `ModSection` refactors) from the actual free mod implementation, this is a bit overwhelming as a single PR.

smoogipoo commented:
Blocked because there's one final issue in that mods and freemods should be compatible - if a mod is chosen the respective incompatible freemods should be deselected and hidden. Awaiting on the mod utils class + the follow up so I'll change gears there.

Also will split out the modselect stuff, though it's all kinda related. I didn't do it because the last times I did this stuff I got blasted for making seemingly unrelated changes with no direction, with PRs building on top of each other.

smoogipoo committed:
Merge branch 'add-mod-utils' into freemods

smoogipoo committed:
Move Mods reset code to OnlinePlaySongSelect

smoogipoo committed:
Merge branch 'add-mod-utils' into freemods

peppy commented:
I think in this case, making a PR with the structural changes to mod select shouldn't feel too "directionless", but I guess we'll see how it looks.

smoogipoo committed:
Disable/disallow freemods on incompatible/selected mods

smoogipoo committed:
Merge branch 'fix-mod-icon' into freemods

smoogipoo committed:
Merge branch 'freemod-select-overlay' into freemods

smoogipoo committed:
Renamespace footer button

smoogipoo committed:
Merge branch 'master' into freemods

smoogipoo committed:
Re-layout to reduce movement

smoogipoo committed:
Fix inspection

smoogipoo committed:
Merge branch 'master' into freemods

smoogipoo commented:
Not sure if the test failure is a regression from this PR or if it's just always been a silent regression and test order has changed.

The test order is:
```
1. TestGlobalModInstancesNotRetained
2. TestNewItemHasNewModInstances
```

1. From test 1, the global `SelectedMods` contains a `DT(2.0)` mod.
2. Test 2 a new default `DT(1.5)` mod is applied.
3. `ModSelectOverlay` binds to the selected mods, and then attempts to transfer the value into the button's mod: https://github.com/smoogipoo/osu/blob/cf5233c6abc693d94dbda84c79fcaea5015eaa56/osu.Game/Overlays/Mods/ModSection.cs#L199-L200
4. The copying of mod settings fails due to an early return for default-value copies: https://github.com/smoogipoo/osu/blob/cf5233c6abc693d94dbda84c79fcaea5015eaa56/osu.Game/Rulesets/Mods/Mod.cs#L151-L152 added in https://github.com/ppy/osu/pull/11386.

frenzibyte commented:
Mostly my fault, previously when this was in `CreateCopy()`, the target mod is instantiated locally and is expected to hold default values, but now that this logic is separated to `CopyFrom`, the target mod could be in a modified state. Will fix.

peppy commented on osu.Game.Tests/Mods/ModValidationTest.cs:
```diff
@@ -0,0 +1,64 @@
+namespace osu.Game.Tests.Mods
+{
+    [TestFixture]
+    public class ModValidationTest
```
Is this class still required? It would seem all these tests now exist in `ModUtilsTest`.

peppy commented on osu.Game/Online/Multiplayer/MultiplayerRoomSettings.cs:
```diff
@@ -32,13 +32,22 @@ public class MultiplayerRoomSettings : IEquatable<MultiplayerRoomSettings>
 
+        [NotNull]
+        [Key(5)]
+        public IEnumerable<APIMod> AllowedMods { get; set; } = Enumerable.Empty<APIMod>();
```
Shall we call this `OptionalMods`? Allowed seems a bit weird since it doesn't include the Mods above.

smoogipoo replied:
I would earlier change the above to `RequiredMods` to match the osu-web API.

peppy replied:
Sounds fine to do that?

smoogipoo replied:
It would break iOS though, due to JSON serialization by name. It shouldn't break messapack though, right?

peppy replied:
ios hasn't even had a release supporting multiplayer, not important at this stage.

smoogipoo replied:
Sure. Do you want that done in this PR?

peppy replied:
Yeah. They can't exist with the current naming pair; feels super weird.

peppy commented on osu.Game/Online/Multiplayer/MultiplayerRoomUser.cs:
```diff
@@ -27,6 +31,13 @@ public class MultiplayerRoomUser : IEquatable<MultiplayerRoomUser>
+        /// </summary>
+        [Key(3)]
+        [NotNull]
+        public IEnumerable<APIMod> UserMods { get; set; } = Enumerable.Empty<APIMod>();
```
Call this `OptionalMods` or just `Mods`? Don't think the user prefix is required when this whole class is applicable to a user.

smoogipoo replied:
Went with `Mods`.

peppy commented:
@smoogipoo thoughts on how we should handle this?

![20210205 123132 (JetBrains Rider)](https://user-images.githubusercontent.com/191335/106986143-1a692180-67ae-11eb-8f6f-a29bd387759f.png)

MessagePack is deserialising to null on a NotNull attributed field.

smoogipoo commented:
Is that happening because the osu-server-spectator isn't updated?

peppy commented:
Correct. I'm testing against an old server version. Are we just going to ignore this one for now as "something which should never happen"?

smoogipoo commented:
I don't really have any idea how to handle that otherwise for now. I'm curious if messagepack would do that if this was marked with non-null (via nullable reference types).

smoogipoo committed:
Rename Mods -> RequiredMods

peppy commented:
Honestly seems like a bug that it can happen in the first place on a `nullable` class? Is that because we haven't enforced nullable on the project, or are there no runtime checks for that stuff?

smoogipoo commented:
Oh that's weird, this is already marked with `#nullable enable`. Definitely feels like a bug then.

smoogipoo committed:
Rename UserMods -> Mods for MultiplayerRoomUser

smoogipoo committed:
Remove ModValidationTest

smoogipoo committed:
Add back vertical spacer

peppy reviewed (approved):
Ignoring test failure for now as unrelated.
